src/actions/dropbox/dropbox.ts

```diff
@@ -0,0 +1,205 @@
+import * as https from "request-promise-native"
+import {URL} from "url"
+
+import Dropbox = require("dropbox")
+import * as winston from "winston"
+import * as Hub from "../../hub"
+
+export class DropboxAction extends Hub.OAuthAction {
+    name = "dropbox"
+    label = "Dropbox"
+    iconName = "dropbox/dropbox.png"
+    description = "Send query results directly to a file in your Dropbox."
+    supportedActionTypes = [Hub.ActionType.Query, Hub.ActionType.Dashboard]
+    usesStreaming = false
+    minimumSupportedLookerVersion = "6.4.0"
+    requiredFields = []
+    params = [
+      {
+        description: "Dropbox Application Key",
+        label: "Application Key",
+        name: "appKey",
+        required: true,
+        sensitive: true,
+      },
+      {
+        description: "Dropbox Secret Key",
+        label: "Secret Key",
+        name: "secretKey",
+        required: true,
+        sensitive: true,
+      },
+    ]
+
+  async execute(request: Hub.ActionRequest) {
+    const filename = request.formParams.filename
+    const directory = request.formParams.directory
+    const ext = request.attachment!.fileExtension
+
+    let accessToken = ""
+    if (request.params.state_json) {
+      const jsonState = JSON.parse(request.params.state_json)
+      if (jsonState.code && jsonState.redirect) {
+        accessToken = await this.getAccessTokenFromCode(request)
+      }
+    }
+    const drop = this.dropboxClientFromRequest(request, accessToken)
+
+    const resp = new Hub.ActionResponse()
+    resp.success = true
+    if (request.attachment && request.attachment.dataBuffer) {
+      const fileBuf = request.attachment.dataBuffer
+      await drop.filesUpload({path: `/${directory}/${filename}.${ext}`, contents: fileBuf}).catch((err: any) => {
+        winston.error(`Upload unsuccessful: ${JSON.stringify(err)}`)
+        resp.success = false
+        resp.state = new Hub.ActionState()
+        resp.state.data = "reset"
+      })
+    } else {
+      resp.success = false
+      resp.message = "No data sent from Looker to be sent to Dropbox"
+    }
+    return resp
+  }
+
+  async form(request: Hub.ActionRequest) {
+    const form = new Hub.ActionForm()
+    form.fields = []
+
+    let accessToken = ""
+    if (request.params.state_json) {
+      try {
+        const jsonState = JSON.parse(request.params.state_json)
+        if (jsonState.code && jsonState.redirect) {
+          accessToken = await this.getAccessTokenFromCode(request)
+        }
+      } catch { winston.warn("Could not parse state_json") }
+    }
+    const drop = this.dropboxClientFromRequest(request, accessToken)
+    try {
+      const response = await drop.filesListFolder({path: ""})
+      let folderList = response.entries.filter((entries) => (entries[".tag"] === "folder"))
+        .map((entries) => ({name: entries.name, label: entries.name}))
+      if (folderList.length === 0) {
+        folderList = [{name: "Looker", label: "Looker"}]
```

Comment:
What is this about?

Reply:
if there are no valid folders to list than do not use "undefined" as the folder name. Actually give a fresh dropbox an option

Reply:
This folder list is a bit wonky then, because it will only show immediate subdirectories of your Dropbox folder, right? I know we don't have a good form option for dealing with hierarchical data. I do think this is a bit confusing though, but I don't know the right answer.